### PR TITLE
TECH 292  - Remove SupportedTypes Constraint

### PR DESCRIPTION
**Pull request by TYRONEMICHAEL**
TECH-292 - Update civic canister to support civic pass
    
    - Remove constraint of supported credential types
    - Remove Credential Error
    - Catch error and return empty credential so the user inspects the JWT in order to see if the credential exists instead of seeing an error when presenting the VC
    - Only return credentials that match the credential request arguments

### CODE REVIEW

**Review by CallForSanity — APPROVED** [review]

File: src/civic_canister_frontend/scripts/issue-credential.ts [hunk 1]
```diff
@@ -29,15 +29,15 @@ const alumniOfClaim = {
 };
 const mixedClaim = {
   claims: [
-    ["Is over 18", { Boolean: true }],
-    ["name", { Text: "Max Mustermann" }],
-    ["alumniOf", { Claim: alumniOfClaim }]
+    ["tigoYhp9SpCDoCQmXGj2im5xa3mnjR1zuXrpCJ5ZRmi", { Boolean: true }],
+    ["Status", { Text: "ACTIVE" }],
+    ["expirationDate", { Number: 1000 + 365 * 24 * 60 * 60 * 1000 }]
```
Comment (CallForSanity) [inline src/civic_canister_frontend/scripts/issue-credential.ts]: I would expect something relative here, e.g. currentData + constant but I guess this is just for demoing anyway?

File: src/civic_canister_backend/src/consent_message.rs [hunk 2]
```diff
@@ -4,38 +4,37 @@ use std::fmt::{Display, Formatter};
...
 use SupportedLanguage::{English, German};
 
-/// Consent messages for the VerifiedAdult VC to be shown and approved to the user during the VC sharing flow 
-const ADULT_VC_DESCRIPTION_EN: &str = r###"# Verified Adult
+/// Consent messages for the CivicPass VC to be shown and approved to the user during the VC sharing flow 
+const VC_DESCRIPTION_EN: &str = r###"# Verifiable Credential
 
-Credential that states that the holder's age is at least 18 years."###;
-const ADULT_VC_DESCRIPTION_DE: &str = r###"# Erwachsene Person
+Credential that states that the holder possesses a Verifiable Credential."###;
```
Comment (dankelleher) [inline src/civic_canister_backend/src/consent_message.rs:16]: Why do we have english and german descriptions?

File: src/civic_canister_frontend/src/App.tsx [hunk 3]
```diff
@@ -63,26 +63,18 @@ function App() {
...
-}
-
 const mixedClaim = {
   claims: [
-    ["Is over 18", { Boolean: true }],
-    ["name", { Text: "Max Mustermann" }],
-    ["alumniOf", { Claim: alumniOfClaim }]
+    ["passType", { Text: "tigoYhp9SpCDoCQmXGj2im5xa3mnjR1zuXrpCJ5ZRmi" }],
+    ["status", { Text: "active" }],
+    ["expirationDate", { Number: 1000 + 365 * 24 * 60 * 60 * 1000 }]
```
Comment (dankelleher) [inline src/civic_canister_frontend/src/App.tsx]: elsewhere (e.g. in your demo vid) this field was a string or date type. Which is correct?

File: src/civic_canister_backend/src/credential.rs [hunk 4]
```diff
@@ -232,10 +214,10 @@ fn is_authorized_issuer(caller: Principal) -> bool {
 async fn add_credentials(
     principal: Principal,
     new_credentials: Vec<Credential>,
-) -> Result<String, CredentialError> {
+) -> Result<String, IssueCredentialError> {
     // Check if the caller is the authorized principal
     if !is_authorized_issuer(caller()) {
-        return Err(CredentialError::UnauthorizedSubject(
+        return Err(IssueCredentialError::UnauthorizedSubject(
```
Comment (dankelleher) [inline src/civic_canister_backend/src/credential.rs]: Should be UnauthorizedIssuer?

File: src/civic_canister_backend/src/credential.rs [hunk 5]
```diff
@@ -280,10 +262,10 @@ async fn add_credentials(
 async fn remove_credential(
     principal: Principal,
     credential_id: String,
-) -> Result<String, CredentialError> {
+) -> Result<String, IssueCredentialError> {
     // Check if the caller is an authorized issuer
     if !is_authorized_issuer(caller()) {
-        return Err(CredentialError::UnauthorizedSubject(
+        return Err(IssueCredentialError::UnauthorizedSubject(
```
Comment (dankelleher) [inline src/civic_canister_backend/src/credential.rs]: Should be RemoveCredentialError and UnauthorizedIssuer?

File: src/civic_canister_backend/src/credential.rs [hunk 6]
```diff
@@ -313,20 +295,20 @@ async fn remove_credential(
                     credentials.insert(principal, CredentialList(existing_credentials_vec));
                     Ok("Credential removed successfully".to_string())
                 } else {
-                    Err(CredentialError::UnauthorizedSubject(
+                    Err(IssueCredentialError::UnauthorizedSubject(
```
Comment (dankelleher) [inline src/civic_canister_backend/src/credential.rs]: Should be RemoveCredentialError and something other than UnauthorizedSubject?

Maybe the enum type should just go back to something more generic like Error or CredentialError if the same values will be used for add/remove/update?

File: src/civic_canister_backend/src/credential.rs [hunk 7]
```diff
@@ -313,20 +295,20 @@ async fn remove_credential(
...
-                Err(CredentialError::NoCredentialFound(format!(
+                Err(IssueCredentialError::UnauthorizedSubject(format!(
                     "Credential not found with id {} for principal {}",
                     credential_id,
                     principal.to_text()
                 )))
             }
         } else {
-            Err(CredentialError::NoCredentialFound(format!(
+            Err(IssueCredentialError::UnauthorizedSubject(format!(
```
Comment (dankelleher) [inline src/civic_canister_backend/src/credential.rs]: Here and above, UnauthorizedSubject seems wrong

File: src/civic_canister_backend/src/consent_message.rs [hunk 8]
```diff
@@ -79,38 +70,27 @@ async fn vc_consent_message(
...
-        Err(err) => {
-            return Err(Icrc21Error::UnsupportedCanisterCall(Icrc21ErrorInfo {
-                description: err,
-            }));
-        }
-    };
     let template = CONSENT_MESSAGE_TEMPLATES
         .get(&(
-            CredentialTemplateType::from(&credential_type),
+            CredentialTemplateType::Credential,
```
Comment (dankelleher) [inline src/civic_canister_backend/src/consent_message.rs:84]: I dont love the idea of having user-readable consent messages hard-coded in the canister. Especially if i18n has to be handled there, as it means each new supported language is a canister change. The canister should be a data store and permissions layer only. Is that something we can delegate to the UI?

File: src/civic_canister_backend/src/consent_message.rs (hunk 2 above)
Comment (dankelleher) [reply src/civic_canister_backend/src/consent_message.rs:16]: Added a related comment below.

Comment (TYRONEMICHAEL) [reply src/civic_canister_backend/src/consent_message.rs:16]: This has been updated

File: src/civic_canister_frontend/src/App.tsx (hunk 3 above)
Comment (TYRONEMICHAEL) [reply src/civic_canister_frontend/src/App.tsx]: This has been updated. It should be a string. Changes were not committed.

File: src/civic_canister_backend/src/credential.rs (hunk 4 above)
Comment (TYRONEMICHAEL) [reply src/civic_canister_backend/src/credential.rs]: Updated

File: src/civic_canister_backend/src/credential.rs (hunk 6 above)
Comment (TYRONEMICHAEL) [reply src/civic_canister_backend/src/credential.rs]: Updated

File: src/civic_canister_backend/src/credential.rs (hunk 7 above)
Comment (TYRONEMICHAEL) [reply src/civic_canister_backend/src/credential.rs]: Updated

File: src/civic_canister_backend/src/consent_message.rs (hunk 8 above)
Comment (TYRONEMICHAEL) [reply src/civic_canister_backend/src/consent_message.rs:84]: As discussed in our meet.

**Review by dankelleher — APPROVED** [review]

**Pull request merged**